src/zkbob/ZkBobPool.sol

```diff
@@ -512,4 +499,39 @@ abstract contract ZkBobPool is IZkBobPool, EIP1967Admin, Ownable, Parameters, Ex
     function _isOwner() internal view override returns (bool) {
         return super._isOwner() || _admin() == _msgSender();
     }
+
+    /**
+     * @dev Appends a commitment to the pending commitments queue.
+     */
+    function _appendCommitment(uint256 _commitment, uint64 _fee, address _prover) internal {
+        pendingCommitments.pushBack(
+            PendingCommitment({commitment: _commitment, fee: _fee, prover: _prover, timestamp: uint64(block.timestamp)})
+        );
+    }
+
+    /**
+     * @dev Validates either the grace period has passed or the caller
+     * is the prover who submitted this commitment.
+     */
+    function _validateGracePeriod(uint64 commitmentTimestamp, address privilegedProver) internal view {
+        // We calculate the beggining of the grace period either from the timestamp of the last tree update,
+        // or from the timestamp of the commitment, whichever is greater.
+        uint64 timestamp = commitmentTimestamp;
+        if (timestamp < lastTreeUpdateTimestamp) {
+            timestamp = lastTreeUpdateTimestamp;
+        }
```

Comment:
So if I submit 100 transactions before someone's else's tx, their effective grace period can be `101 * gracePeriod`? Is this intended?

Reply:
Yes, when you put it this way it doesn't sound good. My original thought was to ensure that each operator had a fair grace period even if other operators handled their commitments slowly or didn't do anything at all. So it seems we have these options:
* Keep the implementation as is. Users might experience some form of DOS, but I don't see any direct economic incentive to do so.
* Remove logic related to lastTreeUpdateTimestamp. Operators might not have a grace period, meaning other operators would have an incentive to send tree update proofs right before the end of the grace period to compete for the next proof.

There probably are some other options, but they might be more complicated. To be honest, I'm also not sure which approach is better.

Reply:
I think timely proof inclusion should be prioritized over operators grace period profitability. 
In the current setting, if all 100 transactions in the queue are sent by different operators, it will become quite a bottleneck and take a lot of time to clear the whole queue of pending tree updates. However, if their grace periods would have expired sooner, single independent operator will be able to clear the whole queue at once more efficiently (even without zk proof batching).
Also, if some operator sees that they won't be able to fit their transaction inside the grace period and worry about potential race condition and transaction reverts, they may choose to not send tree proof at all.

Reply:
Yes, it does make sense. We should prioritize liveness over fairness to the operators. Modified this logic in [d5648f7](https://github.com/zkBob/zkbob-contracts/pull/97/commits/d5648f73b2421878973388eea36374bdf7862453),